### sarek.pathfindr
### Scoring and ranking somatic and germline variations for cancer research

Original code is at https://github.com/NBISweden/pathfindr developed by Markus Mayrhofer. 
Still in progress, but it can have some use. It is intended to run on data generated by [Sarek](https://nf-co.re/sarek), and the goal is to
get filtered, ranked somatic and germline variants, optionally in a HTML page, or in an annotated VCF. This page can help in formulating
a report, or visualize the most important somatic or germline variants. 

To run the code you have to 
  - run Sarek for a normal/tumor sample, get variant calls for ASCAT, Control-FREEC, HaplotypeCaller (calls for both the normal and the tumor sample), Mutect2 (GATK 3.8 for the moment), Strelka and Manta in somatic mode.
  - clone this repo
  - create a conda environment using the provided environment.yml file
  - edit pathfindr.config.yaml
  - run 
    - either `Rscript runPathfindr.R` to have CSV results only
    - or `runPF_html.sh` to get a HTML file
  - after an hour have a look at the results
  
In RStudio console (or R prompt) after checking out from github and changing to the directory you can have 
```
library(devtools)
install("../sarek.pathfindr")
library(sarek.pathfindr)
...
```

To get a HTML output, run `rmarkdown::render("runPF.Rmd")` in the console (or - after installing the proper R5 environment i.e. via conda), do 
```
Rscript <(echo "rmarkdown::render('runPF.Rmd',knit_root_dir='.',output_file='bugger.html')")
```

Csirz: szilveszter.juhos at scilifelab.se
