---
title: "Pathfindr Report"
output:
  html_document:
      toc: false
      toc_float: false
      toc_depth: 4
      code_folding: NULL
      df_print: paged
editor_options:
  chunk_output_type: console
---
<!-- CSS style for left alignment and for tables to look ok: -->
<style>
  body {
    position: absolute;
    left: 42px;
  }
  .main-container {max-width: none !important;}
  table, td, th {
    font-size: 11px;
    border: none;
    padding-left: 1em;
    padding-right: 1em;
    min-width: 100%;
    word-wrap: break-word;
    max-width: 240px;
    margin-left: auto;
    margin-right: auto;
    margin-top: 1em;
    margin-bottom: 1em;
  }
</style>


<!-- Markdown setup -->
```{r setup, include=F}
knitr::opts_chunk$set(echo = FALSE, comment = '', warning=F, error = F)
options(width = 600)
```
# {.tabset .tabset-pills}

```{r project_and_refgenome}
library(config)
library(tictoc)
library(sarek.pathfindr)

cat('Sample:',PFconfig$subject,'\tWorking Directory:',getwd())
cat('Configuration File: pathfindr.config.yaml','\nDate:',date())
```

## Setup {.tabset}

### Package dependencies
```{r package_dependencies}
installed_packages <- as.data.frame(installed.packages()[,c(1,3:4)])
rownames(installed_packages) <- NULL
installed_packages <- installed_packages[is.na(installed_packages$Priority),1:2,drop=FALSE]
print(installed_packages, row.names=FALSE)
```
### List files {.active}
<!-- First, make a list of files to read and visualize -->
```{r list_files}
cat('Reference genome: ',PFconfig$reference)
ascat_header(ascat_files(PFconfig))
freec_header(freec_files(PFconfig))
cat("Mutect2 Somatic Variants file:  ", PFconfig$mutect2file, '\n')
haplotypeCallerFiles <- haplotypeCallerFiles(PFconfig)
cat("HaplotypeCaller Normal file: ", haplotypeCallerFiles[1])
cat("HaplotypeCaller Tumor file: ", haplotypeCallerFiles[2])
manta_files <- mantaFiles(PFconfig)
cat("Manta normal file: ", manta_files["manta_normal_file"][[1]])
cat("Manta tumor file: ", manta_files["manta_tumor_file"][[1]])
cat("Manta allele frequencies file: ... to be added")
strelka_files <- strelkaFiles(PFconfig)
cat("Strelka Somatic SNVs file:  ", strelka_files["strelka_snv_file"], '\n')
cat("Strelka Somatic indels file:  ", strelka_files["strelka_indel_file"], '\n')
```


## Copy number {.tabset}

### Control-FREEC
<!-- Read Control Freec copy number data -->
```{r load_control_freec}
cat("Control-FREEC results")
#freec_results = score_freec(PFconfig)
#tableWrapper(freec_results['FREEC_CNVs'][[1]]) # table in report - I will learn tuples in R one day
```

### Ascat
<!-- Read ASCAT copy number data -->
```{r load_ascat}
cat("ASCAT results")
#ascat_results = score_ascat(PFconfig)
#tableWrapper(ascat_results['ASCAT_CNVs'][[1]]) # table in report
```

### View Control-FREEC

#### Control-FREEC Genome plot
```{r plot_control_freec,fig.width=15,fig.height=7}
cat("Control-FREEC genome plot")
#print( controlFREECGenomeView( freec_results ))
```

&nbsp;

#### Control-FREEC Scatter plot
<!-- Scatter plot -->
```{r scatter_plot_control_freec, echo=FALSE,fig.width=15,fig.height=8,warning=F}
cat("Control-FREEC scatter plot")
#print( controlFREECScatterView( freec_results['binned'][[1]] ) )
```

### View Ascat {.active}

#### ASCAT Genome plot
<!-- Plot ASCAT tumor copy number log ratio, tBAF and nBAF -->
```{r plot_ascat,fig.width=15,fig.height=7}
cat("ASCAT Genome plot")
#print( ascatGenomeView(ascat_results) )
```

&nbsp;

#### ASCAT Scatter plot
```{r scatter_plot_ascat, echo=FALSE,fig.width=15,fig.height=8,warning=F}
cat("ASCAT Scatter plot")
#print( ASCATScatterView( ascat_results['binned'][[1]] ) )
```

## Structural variants {.tabset}

<!-- Read the tumor structural variant data -->
### Manta (tumor)
```{r load_manta_tumor, echo=FALSE}
cat("Manta tumor table")
#manta_vcfs <- mantaFiles(PFconfig)
#somatic_manta <- loadSomaticManta(manta_vcfs)
#tableWrapper(somatic_manta)
```
### Manta (germline)
```{r load_manta_normal, echo=FALSE}
# It takes a while, so skip for testing
cat("Manta Germline table")
# germline_manta <- loadGermlineManta(manta_vcfs)
# tableWrapper(germline_manta)
```

&nbsp;

### View structural variants {.active}
```{r plot_manta, echo=FALSE,fig.width=15,fig.height=4}
  cat("Manta Structural variants plot")
  # we are not printing these for all samples (i.e. for WES there is little point to do so)
  # note, Control-FREEC and ASCAT T ratios,
  # also selected Manta variations are passed
  # in the global 'pfenv' environment
  
  #g <- structuralVariationView(manta_vcfs)
  #if(!is.null(g)) {
  #  print(g)
  #}
```

## Somatic point mutations {.tabset .active}

### Mutect 2
```{r load_mutect2, echo=FALSE}
cat("Mutect2 table")
#mutect2_results <- scoreMutect2(PFconfig)
#tableWrapper(mutect2_results)
```
### Strelka
```{r load_strelka, echo=FALSE}
cat("Strelka table")
#strelka_results <- scoreStrelka(PFconfig)
#tableWrapper(strelka_results)
```

### View Mutect 2
```{r plot_mutect2, echo=FALSE,fig.width=15,fig.height=4}
cat("Mutect2 plot")
#print( mutect2View(mutect2_results) )
```

### View Strelka {.active}
```{r plot_strelka, echo=FALSE,fig.width=15,fig.height=4}
cat("Strelka plot")
#print( strelkaView(strelka_results) )
```

## Germline point mutations {.tabset}

### HaplotypeCaller (normal only)
<!-- Read annotated small variant data (Haplotype caller, both normal and tumor) -->
```{r load_haplotypecaller, echo=FALSE}
haplotypecaller_N_file <- haplotypeCallerFiles(PFconfig)[1]
cat(haplotypecaller_N_file)
haplotypecaller_selected <- scoreHaplotypeCaller(PFconfig)
ix=haplotypecaller_selected$sample==strsplit(basename(haplotypecaller_N_file),'[.]')[[1]][1] # first file is the normal
tableWrapper(haplotypecaller_selected[ix][,-c('cumstart','cumend','DOMAINS')])
```


<!-- ### HaplotypeCaller (all samples) -->

<!-- ### HaplotypeCaller overlap -->
<!-- ```{r haplotypecaller_overlap, echo=FALSE} -->
<!-- cat("Overlap placeholder") -->
<!-- #table <- haplotypeCallerOverlap(haplotypecaller_N_file) -->
<!-- #htmlTable::htmlTable(table) -->
<!-- ``` -->

<!-- ## All {.tabset} -->
<!-- ### Top ranked {.active} -->
<!-- &nbsp; -->
<!-- ### Save to files -->

<!-- ## Chromosomes {.tabset} -->
<!-- ### 1 -->
<!-- ### 2 -->
<!-- ### 3 -->
<!-- ### 4 -->
<!-- ### 5 -->
<!-- ### 6 -->
<!-- ### 7 -->
<!-- ### 8 -->
<!-- ### 9 -->
<!-- ### 10 -->
<!-- ### 11 -->
<!-- ### 12 -->
<!-- ### 13 -->
<!-- ### 14 -->
<!-- ### 15 -->
<!-- ### 16 -->
<!-- ### 17 -->
<!-- ### 18 -->
<!-- ### 19 -->
<!-- ### 20 -->
<!-- ### 21 -->
<!-- ### 22 -->
<!-- ### X -->
<!-- ### Y -->
