---
title: "Pathfindr Report"
output:
  html_document:
      toc: false
      toc_float: false
      toc_depth: 4
      code_folding: NULL
      df_print: paged
editor_options:
  chunk_output_type: console
---
<!-- CSS style for left alignment and for tables to look ok: -->
<style>
  body {
    position: absolute;
    left: 42px;
  }
  .main-container {max-width: none !important;}
  table, td, th {
    font-size: 11px;
    border: none;
    padding-left: 1em;
    padding-right: 1em;
    min-width: 100%;
    word-wrap: break-word;
    max-width: 240px;
    margin-left: auto;
    margin-right: auto;
    margin-top: 1em;
    margin-bottom: 1em;
  }
</style>


<!-- Markdown setup -->
```{r setup, include=F}
knitr::opts_chunk$set(echo = FALSE, comment = '', warning=F, error = F)
options(width = 600)
```
# {.tabset .tabset-pills}
## Setup {.tabset}
### Load dependencies
```{r load_dependencies}
library(config)
library(tictoc)
library(sarek.pathfindr)

makelinks <- function(strings,sep='&') {
  alltumorgenes <- sarek.pathfindr::getEnvVariable("alltumorgenes")
  if (length(strings)>0) for (i in 1:length(strings)) if (strings[i]!='') {
    t=strsplit(strings[i],sep)[[1]]
    for (j in 1:length(t)) {
      if (t[j] %in% alltumorgenes) {
        t[j]=kableExtra::cell_spec(t[j], link = paste0("https://cancer.sanger.ac.uk/cosmic/gene/analysis?ln=",t[j]),format = 'html')
      } else if (startsWith(t[j],'rs')) {
        t[j]=kableExtra::cell_spec(t[j], link = paste0("https://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs=",substr(t[j],3,100)),format = 'html')
      } else if (startsWith(t[j],'COSM')) {
        t[j]=kableExtra::cell_spec(t[j], link = paste0("https://cancer.sanger.ac.uk/cosmic/mutation/overview?id=",substr(t[j],5,100)),format = 'html')
      }
    }
    strings[i]=paste(t,collapse = ' ')
  }
  return(strings)
}

tableWrapper <- function(table) {
  if ('Existing_variation' %in% colnames(table)) {
    table$Existing_variation=makelinks(table$Existing_variation)
  }
  if ('Gene_Name' %in% colnames(table)) {
    table$Gene_Name[nchar(table$Gene_Name)>200]='many'
  }
  if ('SYMBOL' %in% colnames(table)) {
    table$SYMBOL=makelinks(table$SYMBOL)
  }
  if ('cancer_genes' %in% colnames(table)) {
    table$cancer_genes=makelinks(table$cancer_genes,sep=' ')
  }
  if ('AD_TUMOR' %in% colnames(table)) {
    table$AD_TUMOR=unlist(lapply(table$AD_TUMOR,paste,collapse=', '))
  }
  if ('AD_NORMAL' %in% colnames(table)) {
    table$AD_NORMAL=unlist(lapply(table$AD_NORMAL,paste,collapse=', '))
  }
  if ('AD' %in% colnames(table)) {
    if (is.list(table$AD)) table$AD=unlist(lapply(table$AD,paste,collapse=', '))
  }

  samples=sort(unique(table$sample))

  if (T) if (nrow(table)>0) htmlTable::htmlTable(table,
            col.rgroup = c("none", "#F7F7F7"),
            tspanner=paste('Rank score:',unique(table$rank_score)),
            n.tspanner=rev(table(table$rank_score)))
  # if (length(samples)>1) htmlTable(table, 
  #           col.rgroup = c("none", "#F7F7F7"),
  #           n.rgroup='',
  #           tspanner=paste('Rank score:',unique(table$rank_score)),
  #           n.tspanner=table(table$rank_score))
}
```

## Copy number {.tabset}

#### ControlFreec
<!-- Read Control Freec copy number data -->
```{r load_control_freec}
tic("Control-FREEC scores and LOH")
freec_results = score_freec(PFconfig)
toc()
tableWrapper(freec_results['FREEC_CNVs'][[1]]) # table in report - I will learn tuples in R one day
```
#### Control-FREEC Genome plot
```{r plot_control_freec,fig.width=15,fig.height=7}
print( controlFREECGenomeView( freec_results ))
```
&nbsp;
#### Control-FREEC Scatter plot
<!-- Scatter plot -->
```{r scatter_plot_control_freec, echo=FALSE,fig.width=15,fig.height=8,warning=F}
print( controlFREECScatterView( freec_results['binned'][[1]] ) )
```
<!-- ### View Ascat {.active} -->
<!-- #### Ascat -->
<!-- <!-- Read ASCAT copy number data --> -->
<!-- ```{r load_ascat} -->
<!-- tic("ASCAT scores and LOH") -->
<!-- ascat_results = score_ascat(PFconfig) -->
<!-- toc() -->
<!-- tableWrapper(ascat_results['ASCAT_CNVs'][[1]]) # table in report -->
<!-- ``` -->
<!-- #### ASCAT Genome plot -->
<!-- <!-- Plot ASCAT tumor copy number log ratio, tBAF and nBAF --> -->
<!-- ```{r plot_ascat,fig.width=15,fig.height=7} -->
<!-- print( ascatGenomeView(ascat_results) ) -->
<!-- ``` -->
<!-- &nbsp; -->
<!-- #### ASCAT Scatter plot -->
<!-- ```{r scatter_plot_ascat, echo=FALSE,fig.width=15,fig.height=8,warning=F} -->
<!-- print( ASCATScatterView( ascat_results['binned'][[1]] ) ) -->
<!-- ``` -->

## Structural variants {.tabset}

<!-- Read the tumor structural variant data -->
### Manta (tumor)
```{r load_manta_tumor, echo=FALSE}
manta_vcfs <- mantaFiles(PFconfig)
somatic_manta <- loadSomaticManta(manta_vcfs)
tableWrapper(somatic_manta)
```
### Manta (germline)
```{r load_manta_normal, echo=TRUE}
# It takes a while, so skip for testing
print("Manta Germline Results should be here")
# germline_manta <- loadGermlineManta(manta_vcfs)
# tableWrapper(germline_manta)
```

&nbsp;

### View structural variants {.active}
```{r plot_manta, echo=FALSE,fig.width=15,fig.height=4}
  # we are not printing these for all samples (i.e. for WES there is little point to do so)
  # note, Control-FREEC and ASCAT T ratios,
  # also selected Manta variations are passed
  # in the global 'pfenv' environment
  g <- structuralVariationView(manta_vcfs)
  if(!is.null(g)) {
    print(g)
  }
```

## Somatic point mutations {.tabset .active}

### Mutect 2
```{r load_mutect2, echo=FALSE}
mutect2_results <- scoreMutect2(PFconfig)
tableWrapper(mutect2_results)
```
### Strelka
```{r load_strelka, echo=FALSE}
strelka_results <- scoreStrelka(PFconfig)
tableWrapper(strelka_results)
```

### View Mutect 2
```{r plot_mutect2, echo=FALSE,fig.width=15,fig.height=4}
print( mutect2View(mutect2_results) )
```

### View Strelka {.active}
```{r plot_strelka, echo=FALSE,fig.width=15,fig.height=4}
print( strelkaView(strelka_results) )
```

## Germline point mutations {.tabset}
### HaplotypeCaller (normal only)
### HaplotypeCaller (all samples)
### HaplotypeCaller overlap

## All {.tabset}
### Top ranked {.active}
&nbsp;
### Save to files

## Chromosomes {.tabset}
### 1
### 2
### 3
### 4
### 5
### 6
### 7
### 8
### 9
### 10
### 11
### 12
### 13
### 14
### 15
### 16
### 17
### 18
### 19
### 20
### 21
### 22
### X
### Y
